File: r/basics/quicksort.R
# Sort a fixed vector with a quicksort over the head and tail of the vector.
# Run: Rscript quicksort.R

quicksort <- function(items) {
  if (length(items) <= 1) {
    return(items)
  }

  pivot <- items[1]
  rest <- items[-1]
  c(quicksort(rest[rest <= pivot]), pivot, quicksort(rest[rest > pivot]))
}

cat(paste(quicksort(c(5, 3, 8, 4, 2, 7, 1, 10, 9, 6)), collapse = " "), "\n", sep = "")

File: ruby/basics/quicksort.rb
#!/usr/bin/env ruby
# Sort a fixed array with a quicksort over the head and tail of the array.

def quicksort(items)
  return items if items.size <= 1

  pivot, *rest = items
  quicksort(rest.select { |x| x <= pivot }) + [pivot] + quicksort(rest.select { |x| x > pivot })
end

puts quicksort([5, 3, 8, 4, 2, 7, 1, 10, 9, 6]).join(' ')
